File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/util/MapUtils.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.trace.core.util;

import com.google.gson.JsonObject;

import java.util.Map;
import java.util.Objects;
import com.datadog.android.trace.internal.compat.function.Function;

public class MapUtils {

    public static <V, K> V computeIfAbsent(K key, Map<K, V> originalMap,
                                           Function<? super K, ? extends V> mappingFunction) {
        Objects.requireNonNull(mappingFunction);
        V v;
        if ((v = originalMap.get(key)) == null) {
            V newValue;
            if ((newValue = mappingFunction.apply(key)) != null) {
                originalMap.put(key, newValue);
                return newValue;
            }
        }

        return v;
    }

    public static JsonObject getAsJsonObject(Map<String, String> map) {
        final JsonObject jsonObject = new JsonObject();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            jsonObject.addProperty(entry.getKey(), entry.getValue());
        }
        return jsonObject;
    }

}

File: dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/utils/MapUtils.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.android.core.internal.utils

/**
 * @return a new mutable map containing all key-value pairs from the given array of pairs.
 *
 * The returned map preserves the entry iteration order of the original array.
 * If any of two pairs would have the same key the last one gets added to the map.
 */
internal fun <K, V> Iterable<Pair<K, V>>.toMutableMap(): MutableMap<K, V> {
    return toMap(mutableMapOf())
}

/**
 * @return the [MutableMap] if its not `null`, or the empty [MutableMap] otherwise.
 */
internal fun <K, V> MutableMap<K, V>?.orEmpty(): MutableMap<K, V> {
    return this ?: mutableMapOf()
}
